File: reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        
        if(head == null || head.next == null){
            return;
        }
        
        ListNode fast = head;
        ListNode slow = head;
        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        
        Stack<ListNode> stack = new Stack<>();
        
        fast = slow;
        while(fast != null){
            stack.push(fast);
            fast = fast.next;
        }
 
        slow = head;
        ListNode nex = null;
        while(slow != null){
            nex = slow.next;
            if(stack.isEmpty()){
                slow = nex;
                break;
            }
            slow.next = stack.pop();
            slow = slow.next;
            slow.next = nex;
            slow = nex;
        }
        slow.next = null;
    }
}

File: reorder-list/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        stack<ListNode*> st;
        ListNode* cur = head;
        while(cur != NULL){
            st.push(cur);
            cur = cur->next;
        }
        ListNode* pass = head->next;
        cur = head;
        while(!st.empty() && cur != pass && cur != st.top()){
            cur->next = st.top();
            st.pop();
            cur = cur->next;
            cur->next = NULL;
            cur->next = pass;
            cur = pass;
            pass = pass->next;
            cur->next = NULL;
        }
        if(cur == pass){
            cur->next = pass;
            pass->next = NULL;
        }
        else if(cur == st.top()){
            cur->next = st.top();
            cur = cur->next;
            cur->next = NULL;
        }
    }
};